File: app/src/main/java/com/easysteps/viewmodel/LoginViewModel.java
package com.easysteps.viewmodel;

import android.content.Context;
import android.widget.Toast;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.easysteps.helper.Utils;
import com.easysteps.network.ApiResponse;
import com.easysteps.repository.Repository;

import java.util.HashMap;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by NIKUNJ
 */

public class LoginViewModel extends ViewModel {

    private final CompositeDisposable disposables = new CompositeDisposable();
    private final MutableLiveData<ApiResponse> responseLiveData = new MutableLiveData<>();
    private final Repository repository;

    public LoginViewModel(Repository repository) {
        this.repository = repository;
    }

    public MutableLiveData<ApiResponse> loginResponse() {
        return responseLiveData;
    }

    public void getLogin(Context context, HashMap<String, String> body) {
        if (Utils.checkInternetConnection(context))
            disposables.add(repository.login(body)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnSubscribe((d) -> responseLiveData.setValue(ApiResponse.loading()))
                    .subscribe(
                            result -> responseLiveData.setValue(ApiResponse.success(result)),
                            throwable -> responseLiveData.setValue(ApiResponse.error(throwable))
                    ));
        else Toast.makeText(context, "No internet connection", Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onCleared() {
        disposables.clear();
    }
}

File: app/src/main/java/com/easysteps/viewModel/LoginViewModel.kt
package com.easysteps.viewModel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.easysteps.base.BaseViewModel
import com.easysteps.base.SingleLiveEvent
import com.easysteps.viewModel.models.Login
import kotlinx.coroutines.launch

/**
 * Created by NIKUNJ on 29-04-2022.
 */
class LoginViewModel : BaseViewModel() {

    private val _loginData = SingleLiveEvent<Login>()
    val loginData: LiveData<Login> = _loginData

    private val _loginError = MutableLiveData<Throwable>()
    val dataError: LiveData<Throwable> = _loginError

    fun callApi(map: HashMap<String, Any>) {
        viewModelScope.launch {
            displayLoader()
            processDataEvent(api1Repository.login(map), onError = {
                _loginError.postValue(it)
            }) {
                _loginData.postValue(it)
            }
        }
    }

}